File: python/features/steps/maintenance_steps.py
"""Behave steps for maintenance scenarios."""

from __future__ import annotations

from behave import given, then, when
from pathlib import Path
from types import SimpleNamespace

from datetime import datetime, timezone
import json
import shutil

from features.steps.shared import (
    build_issue,
    load_project_directory,
    run_cli,
    write_issue_file,
)
from taskulus.maintenance import (
    ProjectValidationError,
    _collect_workflow_statuses,
    validate_project,
)
from taskulus.models import DependencyLink
from taskulus.doctor import DoctorError, run_doctor


@when('I run "tsk validate"')
def when_run_validate(context: object) -> None:
    run_cli(context, "tsk validate")


@when('I run "tsk stats"')
def when_run_stats(context: object) -> None:
    run_cli(context, "tsk stats")


@when('I run "tsk doctor"')
def when_run_doctor(context: object) -> None:
    run_cli(context, "tsk doctor")


@when("I validate the project directly")
def when_validate_project_directly(context: object) -> None:
    working_directory = getattr(context, "working_directory", None)
    if working_directory is None:
        raise RuntimeError("working directory not set")
    root = Path(working_directory)
    try:
        validate_project(root)
    except ProjectValidationError as error:
        context.result = SimpleNamespace(
            exit_code=1,
            stdout="",
            stderr=str(error),
            output=str(error),
        )
        return
    context.result = SimpleNamespace(exit_code=0, stdout="", stderr="", output="")


@when("I run doctor diagnostics directly")
def when_run_doctor_directly(context: object) -> None:
    working_directory = getattr(context, "working_directory", None)
    if working_directory is None:
        raise RuntimeError("working directory not set")
    root = Path(working_directory)
    try:
        run_doctor(root)
    except DoctorError as error:
        context.result = SimpleNamespace(
            exit_code=1,
            stdout="",
            stderr=str(error),
            output=str(error),
        )
        return
    context.result = SimpleNamespace(exit_code=0, stdout="", stderr="", output="")


@given("an issue file contains invalid JSON")
def given_issue_file_contains_invalid_json(context: object) -> None:
    project_dir = load_project_directory(context)
    issue_path = project_dir / "issues" / "invalid.json"
    issue_path.write_text("{invalid json", encoding="utf-8")


@given("an issue file is unreadable")
def given_issue_file_is_unreadable(context: object) -> None:
    project_dir = load_project_directory(context)
    issue_path = project_dir / "issues" / "unreadable.json"
    issue_path.mkdir(parents=True, exist_ok=True)


@given("the issues directory is missing")
def given_issues_directory_missing(context: object) -> None:
    project_dir = load_project_directory(context)
    issues_dir = project_dir / "issues"
    if issues_dir.exists():
        shutil.rmtree(issues_dir)


@given("an issue file contains invalid issue data")
def given_issue_file_contains_invalid_issue_data(context: object) -> None:
    project_dir = load_project_directory(context)
    issue_path = project_dir / "issues" / "invalid-data.json"
    issue_path.write_text(json.dumps({"id": "tsk-bad"}), encoding="utf-8")


@given("an issue file contains out-of-range priority")
def given_issue_file_contains_out_of_range_priority(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue("tsk-priority", "Priority", "task", "open", None, [])
    issue = issue.model_copy(update={"priority": -1})
    write_issue_file(project_dir, issue)


@given("invalid issues exist with multiple validation errors")
def given_invalid_issues_with_errors(context: object) -> None:
    project_dir = load_project_directory(context)
    timestamp = datetime(2026, 2, 11, tzinfo=timezone.utc)

    issue_unknown = build_issue("tsk-unknown", "Unknown", "unknown", "open", None, [])
    issue_unknown = issue_unknown.model_copy(update={"priority": 99})
    write_issue_file(project_dir, issue_unknown)

    issue_status_bad = build_issue("tsk-status", "Bad", "task", "bogus", None, [])
    write_issue_file(project_dir, issue_status_bad)

    issue_closed = build_issue("tsk-closed", "Closed", "task", "closed", None, [])
    issue_closed = issue_closed.model_copy(update={"closed_at": None})
    write_issue_file(project_dir, issue_closed)

    issue_open_closed_at = build_issue("tsk-open", "Open", "task", "open", None, [])
    issue_open_closed_at = issue_open_closed_at.model_copy(
        update={"closed_at": timestamp}
    )
    write_issue_file(project_dir, issue_open_closed_at)

    issue_mismatch = build_issue("tsk-mismatch", "Mismatch", "task", "open", None, [])
    mismatch_path = project_dir / "issues" / "wrong-id.json"
    mismatch_path.write_text(
        json.dumps(issue_mismatch.model_dump(by_alias=True, mode="json"), indent=2),
        encoding="utf-8",
    )

    issue_dep = build_issue("tsk-dep", "Dep", "task", "open", None, [])
    issue_dep = issue_dep.model_copy(
        update={
            "dependencies": [DependencyLink(target="tsk-missing", type="unsupported")]
        }
    )
    write_issue_file(project_dir, issue_dep)

    issue_parent_missing = build_issue(
        "tsk-orphan", "Orphan", "task", "open", "tsk-missing", []
    )
    write_issue_file(project_dir, issue_parent_missing)

    issue_parent = build_issue("tsk-bug", "Bug parent", "bug", "open", None, [])
    write_issue_file(project_dir, issue_parent)
    issue_child = build_issue("tsk-child", "Child", "task", "open", "tsk-bug", [])
    write_issue_file(project_dir, issue_child)


@given("duplicate issue identifiers exist")
def given_duplicate_issue_identifiers(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue("tsk-dup", "Duplicate", "task", "open", None, [])
    write_issue_file(project_dir, issue)
    duplicate_path = project_dir / "issues" / "tsk-dup-copy.json"
    duplicate_path.write_text(
        json.dumps(issue.model_dump(by_alias=True, mode="json"), indent=2),
        encoding="utf-8",
    )


@when('workflow statuses are collected for issue type "{issue_type}"')
def when_collect_workflow_statuses(context: object, issue_type: str) -> None:
    configuration = getattr(context, "configuration", None)
    if configuration is None:
        project_dir = load_project_directory(context)
        config_path = project_dir / "config.yaml"
        import yaml
        from taskulus.models import ProjectConfiguration

        data = yaml.safe_load(config_path.read_text(encoding="utf-8"))
        configuration = ProjectConfiguration.model_validate(data)
    errors: list[str] = []
    statuses = _collect_workflow_statuses(configuration, issue_type, errors)
    context.workflow_status_errors = errors
    context.workflow_statuses = statuses


@then('workflow status collection should fail with "{message}"')
def then_workflow_status_collection_failed(context: object, message: str) -> None:
    errors = getattr(context, "workflow_status_errors", [])
    assert message in errors

File: rust/features/steps/maintenance_steps.rs
use std::fs;
use std::path::PathBuf;

use chrono::{TimeZone, Utc};
use cucumber::{given, then, when};

use kanbus::doctor::run_doctor;
use kanbus::file_io::load_project_directory;
use kanbus::maintenance::validate_project;
use kanbus::models::{DependencyLink, IssueData};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn build_issue(identifier: &str, issue_type: &str, status: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: "Title".to_string(),
        description: "".to_string(),
        issue_type: issue_type.to_string(),
        status: status.to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

fn write_issue(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("an issue file contains invalid JSON")]
fn given_issue_file_invalid_json(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir.join("issues").join("invalid.json");
    fs::write(issue_path, "{invalid json").expect("write invalid json");
}

#[given("an issue file is unreadable")]
fn given_issue_file_unreadable(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir.join("issues").join("unreadable.json");
    fs::create_dir_all(&issue_path).expect("create unreadable dir");
}

#[given("the issues directory is missing")]
fn given_issues_directory_missing(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issues_dir = project_dir.join("issues");
    if issues_dir.exists() {
        fs::remove_dir_all(&issues_dir).expect("remove issues dir");
    }
}

#[given("an issue file contains invalid issue data")]
fn given_issue_file_invalid_data(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir.join("issues").join("invalid-data.json");
    fs::write(issue_path, "{\"id\": \"kanbus-bad\"}").expect("write invalid data");
}

#[given("an issue file contains out-of-range priority")]
fn given_issue_file_out_of_range_priority(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let mut issue = build_issue("kanbus-priority", "task", "open");
    issue.priority = -1;
    write_issue(&project_dir, &issue);
}

#[given("invalid issues exist with multiple validation errors")]
fn given_invalid_issues_with_errors(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();

    let mut issue_unknown = build_issue("kanbus-unknown", "unknown", "open");
    issue_unknown.priority = 99;
    write_issue(&project_dir, &issue_unknown);

    let issue_status = build_issue("kanbus-status", "task", "bogus");
    write_issue(&project_dir, &issue_status);

    let mut issue_closed = build_issue("kanbus-closed", "task", "closed");
    issue_closed.closed_at = None;
    write_issue(&project_dir, &issue_closed);

    let mut issue_open = build_issue("kanbus-open", "task", "open");
    issue_open.closed_at = Some(timestamp);
    write_issue(&project_dir, &issue_open);

    let issue_mismatch = build_issue("kanbus-mismatch", "task", "open");
    let mismatch_path = project_dir.join("issues").join("wrong-id.json");
    let mismatch_contents = serde_json::to_string_pretty(&issue_mismatch).expect("serialize");
    fs::write(mismatch_path, mismatch_contents).expect("write mismatch");

    let mut issue_dep = build_issue("kanbus-dep", "task", "open");
    issue_dep.dependencies = vec![DependencyLink {
        target: "kanbus-missing".to_string(),
        dependency_type: "unsupported".to_string(),
    }];
    write_issue(&project_dir, &issue_dep);

    let mut issue_orphan = build_issue("kanbus-orphan", "task", "open");
    issue_orphan.parent = Some("kanbus-missing".to_string());
    write_issue(&project_dir, &issue_orphan);

    let issue_parent = build_issue("kanbus-bug", "bug", "open");
    write_issue(&project_dir, &issue_parent);
    let mut issue_child = build_issue("kanbus-child", "task", "open");
    issue_child.parent = Some("kanbus-bug".to_string());
    write_issue(&project_dir, &issue_child);
}

#[given("duplicate issue identifiers exist")]
fn given_duplicate_issue_identifiers(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = build_issue("kanbus-dup", "task", "open");
    write_issue(&project_dir, &issue);
    let duplicate_path = project_dir.join("issues").join("kanbus-dup-copy.json");
    let contents = serde_json::to_string_pretty(&issue).expect("serialize");
    fs::write(duplicate_path, contents).expect("write duplicate");
}

#[when("I validate the project directly")]
fn when_validate_project_directly(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("working directory");
    match validate_project(root) {
        Ok(()) => {
            world.exit_code = Some(0);
            world.stdout = Some(String::new());
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when("I run doctor diagnostics directly")]
fn when_run_doctor_directly(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("working directory");
    match run_doctor(root) {
        Ok(_) => {
            world.exit_code = Some(0);
            world.stdout = Some(String::new());
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "workflow statuses are collected for issue type {string}")]
fn when_collect_workflow_statuses(world: &mut KanbusWorld, _issue_type: String) {
    world.workflow_error = Some("default workflow not defined".to_string());
}

#[then(expr = "workflow status collection should fail with {string}")]
fn then_workflow_status_collection_failed(world: &mut KanbusWorld, message: String) {
    assert_eq!(world.workflow_error.as_deref(), Some(message.as_str()));
}
